plugin/src/main/java/io/jenkins/plugins/casc/impl/secrets/VaultSecretSource.java

```diff
@@ -105,22 +108,28 @@ private void determineAuthenticator(Properties prop) {
         Optional<String> vaultAppRoleSecret = getVariable(CASC_VAULT_APPROLE_SECRET, prop);
 
         vaultToken.ifPresent(
-                token -> vaultAuthenticator = new VaultSingleTokenAuthenticator(token)
+                token -> configuredAuthenticator = new VaultSingleTokenAuthenticator(token)
         );
 
         vaultUser.ifPresent(
                 user -> vaultPw.ifPresent(
-                        pw -> vaultAuthenticator = new VaultUserPassAuthenticator(user, pw, vaultMount.orElse(DEFAULT_USER_BACKEND))
+                        pw -> configuredAuthenticator = new VaultUserPassAuthenticator(user, pw, vaultMount.orElse(DEFAULT_USER_BACKEND))
                 )
         );
 
         vaultAppRole.ifPresent(
                 approle -> vaultAppRoleSecret.ifPresent(
-                        approleSecret -> vaultAuthenticator = new VaultAppRoleAuthenticator(approle, approleSecret)
+                        approleSecret -> configuredAuthenticator = new VaultAppRoleAuthenticator(approle, approleSecret)
                 )
         );
 
-        if (vaultAuthenticator == null) {
+        // Overwrite current authenticator only if there was a change, because we do not want to loose current auth token
+        if (configuredAuthenticator != null
+                && !configuredAuthenticator.equals(currentlyUsedAuthenticator)) {
+            currentlyUsedAuthenticator = configuredAuthenticator;
```

Comment:
Move to a method, avoiding the need for two VaultAuthenticator fields and cleans up lambdas too. You could implement the new `YamlSource.of` Method and reduce it even more to an actual method reference.

https://github.com/jenkinsci/configuration-as-code-plugin/blob/f16aebb28d5c29aeb9d9582d05e8772d92708d26/plugin/src/main/java/io/jenkins/plugins/casc/yaml/YamlSource.java#L25-L48

Reply:
I started implementing, cause I wanted to see the code :smile:

Reply:
@fishi0x01 What do you think of my additions? 😅

Reply:
Yes looks awesome! Thank you :)